### PR TITLE
GH-41464: [Python] Fix StructArray.sort() for by=None

### PR DESCRIPTION
**Pull request by a-reich**
### Rationale for this change
Closes issue https://github.com/apache/arrow/issues/41464. Fix `StructArray.sort` method's `by` param to work in the case of `by=None` which was documented to mean sort by all fields (the default), but would raise an exception.

### What changes are included in this PR?
* Add a unit test with by=None in `test_struct_array_sort` that fails on main
* Fix the sort method

### Are these changes tested?
yes

### Are there any user-facing changes?
yes
* GitHub Issue: #41464

### CODE REVIEW

**Comment by a-reich** [comment]
Can someone give this small change a review? Perhaps @jorisvandenbossche since you responded to the issue? I appreciate all your efforts.

**Comment by jorisvandenbossche** [comment]
> Hmm, the GHA failure seems completely unrelated to this change - looks like it’s a known flaky test problem #40675.

That's correct, that's a known flaky test. You can just ignore it for now (all other builds are passing, so that is sufficient for this PR)

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -3925,7 +3925,8 @@ cdef class StructArray(Array):
             tosort = self
         indices = _pc().sort_indices(
             tosort,
-            options=_pc().SortOptions(sort_keys=[("", order)], **kwargs)
+            options=_pc().SortOptions(
+                sort_keys=[(field.name, order) for field in self.type], **kwargs)
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:3929]: We should probably define this `sort_keys` also in the `if`/`else` clause above, so we can keep the `sort_keys=[("", order)]` for the case `by` was specified. 
This doesn't cause any issue right now, but that feels safer if we would be stricter in the future in `sort_indices` for a flat array, that we would be sorting it by multiple keys.

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot! Added a small suggestion

File: python/pyarrow/array.pxi (hunk 1 above)
Comment (a-reich) [reply python/pyarrow/array.pxi]: I think I see what you mean, thanks! Pushed https://github.com/apache/arrow/pull/41495/commits/52b9d2c8cf7e6c16adee9246e30d01e1a1715700.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good, thanks!

**Pull request merged**